src/main/java/com/kodilla/ecommercee/service/GroupService.java

```diff
@@ -29,5 +31,28 @@ public void deleteGroup(final Long id) {
         groupRepository.deleteById(id);
     }
 
+    public Optional<Group> getGroupByName(final String name) {
+        return groupRepository.findGroupByName(name);
+    }
 
+    public void deleteGroupAndMoveProductToUnassignedGroup(long id) {
+        Optional<Group> unassignedGroup = getGroupByName("Unassigned product.");
+        Optional<Group> optionalGroup = getGroup(id);
+        if(!unassignedGroup.isPresent()) {
+            Group newUnassignedGroup = new Group("Unassigned product.");
+            saveGroup(newUnassignedGroup);
+            optionalGroup.ifPresent(group -> group.getProducts().forEach(p -> p.setGroup(newUnassignedGroup)));
+            optionalGroup.ifPresent(group -> group.getProducts().forEach(p -> productService.saveProduct(p)));
```

Comment:
4 pętle for, w tym w jednej cały czas leci save do bazy danych. Optymalnie by było przygotować listę obiektów jakie chcesz zapisać w bazie danych i je po prostu zapisać. Zamiast X zapytań będzie 1

Reply:
zapisanie  10 k  obiektów osobno w pętli  zajęło 30s
zapisanie tych samych obiektów  jako listy 3s

faktycznie różnica ogromna!  teraz to zapamiętam  :)